crit/mempages_test.go

```diff
@@ -0,0 +1,123 @@
+package crit
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"regexp"
+	"strconv"
+	"strings"
+	"testing"
+)
+
+const (
+	testImgsDir = "test-imgs"
+)
+
+// TestNewMemoryReader tests the NewMemoryReader function.
+func TestNewMemoryReader(t *testing.T) {
+	pid, err := getTestImgsPID(testImgsDir)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testCases := []struct {
+		name          string
+		dir           string
+		pid           int
+		expectedError error
+	}{
+		{
+			name:          "Valid test-imgs directory and pid",
+			dir:           testImgsDir,
+			pid:           pid,
+			expectedError: nil,
+		},
+		{
+			name:          "Invalid test-imgs directory",
+			dir:           "no test directory",
+			pid:           pid,
+			expectedError: errors.New("no such file or directory"),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			mr, err := NewMemoryReader(tc.dir, tc.pid)
+			if err != nil && !strings.Contains(err.Error(), tc.expectedError.Error()) {
+				t.Errorf("Expected error: %v, got error: %v", tc.expectedError, err)
+			}
+
+			if mr == nil && tc.expectedError == nil {
+				t.Errorf("MemoryReader creation failed for checkpoint directory: %s and pid: %d", tc.dir, tc.pid)
+			}
+		})
+	}
+}
+
+func TestGetMemPages(t *testing.T) {
+	pid, err := getTestImgsPID(testImgsDir)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	type testcase struct {
+		name          string
+		start         uint64
+		end           uint64
+		expectedError error
+	}
+
+	testCases := []testcase{
+		{
+			name:          "Zero memory area size",
+			start:         0,
+			end:           0,
+			expectedError: errors.New("memory size must be greater than 0"),
+		},
+	}
+
+	mr, err := NewMemoryReader(testImgsDir, pid)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for i, pm := range mr.pagemapEntries[:2] {
+		testCases = append(testCases, testcase{
+			name:          fmt.Sprintf("Valid pagemap address %d", i+1),
+			start:         pm.GetVaddr(),
+			end:           pm.GetVaddr() + uint64(uint32(pageSize)*pm.GetNrPages()),
+			expectedError: nil,
+		})
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			buff, err := mr.GetMemPages(tc.start, tc.end)
+			if err != nil && !strings.Contains(err.Error(), tc.expectedError.Error()) {
+				t.Errorf("Expected error: %v, got error: %v", tc.expectedError, err)
+			}
+
+			if tc.expectedError == nil && buff.Len() < 1 {
+				t.Errorf("Returned memory chunk is expected to be non-empty")
+			}
+		})
+	}
+}
+
+func getTestImgsPID(dir string) (int, error) {
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return 0, fmt.Errorf("Failed to read directory: %w\n", err)
+	}
+
+	re := regexp.MustCompile(`pagemap-(\d+)\.img`)
```

Comment:
You should be able to figure out the PID using other images. Not by parsing one of the file names.